### PR TITLE
Fix AutoTP + deep compile collectives silently drop when AC is on

### LINKED ISSUE #8357: [BUG]AutoTP compile pass silently dropping tensor-parallel collectives
# [DeepCompile] Fix AutoTP compile pass silently dropping tensor-parallel collectives

## Summary

The AutoTP compile pass can leave a tensor-parallel layer with **no collective at all** — neither
the module-level one it switched off, nor a graph-level replacement. The run does not fail. It
produces wrong forward values and wrong gradients, and exits 0.

Any model using activation checkpointing hits this, which includes anything calling
HuggingFace's `model.gradient_checkpointing_enable()`.

## The bug

`defer_collectives_to_compiler` switches off the module-level collective on every injected
`TensorParallel_Layer`, on the promise that the graph pass will emit a replacement:

```python
for module in tp_modules:
    module.defer_collectives_to_compiler = True   # module-level collective now off
```

`pass_insert_tp_collectives` then walks `gm.graph.nodes` and inserts `f`/`g` for the parallel
matmuls it finds. **Nothing checks that the two sets agree.**

Under activation checkpointing they do not. Dynamo lifts a checkpointed region into a child
`GraphModule` behind a `tag_activation_checkpoint` node, and the pass never descends into it. The
layers inside keep neither collective, so a row-parallel matmul's output stays a partial sum.

`fullgraph=True` does not protect against this — the higher-order op *is* a full-graph capture, so
nothing graph-breaks and nothing warns.

### Why this is not a niche configuration

HuggingFace's `gradient_checkpointing_enable()` defaults to
`gradient_checkpointing_kwargs={"use_reentrant": False}` (`transformers/modeling_utils.py`), and
`use_reentrant=False` is exactly the variant Dynamo captures into a subgraph rather than
graph-breaking on. So `model.gradient_checkpointing_enable()` plus `"passes": ["autotp"]` silently
corrupts training.

### Measured

2×H200, tp=2, compared against the module-level AutoTP path — the same reference the pass's own
tests use:

| model | forward / loss max diff | worst grad diff | collectives inserted | parallel matmuls hidden in subgraphs |
|---|---|---|---|---|
| MLP, no checkpointing | 2.4e-07 | 7.2e-07 | 6 of 6 | 0 |
| MLP, checkpointing | **2.7e-01** | **9.2e-01** | 2 of 8 | 6 |
| HF Llama + `gradient_checkpointing_enable()` | **3.1e-02** | **2.6e-01** | **0 of 14** | **14** |

Exit code 0 in every row.

Worth noting the middle row: insertion is *partial*, not absent. Layers outside a checkpointed
block still receive their collectives, so the graph looks instrumented on inspection.

### PR DESCRIPTION
**Pull request by pengdurice**
Fix https://github.com/deepspeedai/DeepSpeed/issues/8357
## The fix

Two independent changes.

**1. Walk every graph, then verify the result.**

`iter_graphs()` yields the top-level graph and every nested `GraphModule`, so checkpointed regions
are covered. The pass then compares the tensor-parallel layers it *reached* in the graph against the
ones it gave a collective to, and raises if any are missing.

The check matters more than the traversal fix. It catches the next variant of this failure — a
matmul traced to an op `_MATMUL_TARGETS` does not list, a future change to `nn_module_stack`,
another higher-order op — instead of only the checkpointing case fixed here. Given the failure mode
is silent wrong numbers, a loud error is the right default.

The check is fail-closed. The filter against the deferred-layer names is applied only when the two
naming schemes demonstrably overlap, because intersecting mismatched names would quietly empty the
set and turn the check into a no-op — the exact failure it exists to catch.

**2. Fail if a deferred layer runs eagerly.**

`defer_collectives_to_compiler` is set during `engine.compile()`, but `nn.Module.compile()` is lazy:
compilation happens at the first forward. With `TORCH_COMPILE_DISABLE=1` or
`torch._dynamo.config.disable` the backend never runs, the model executes eagerly forever with its
collectives switched off, and training is silently wrong. The affected layers now assert they are
inside a compiled region.

## Tests

| Test | Covers |
|---|---|
| `TestAutoTPCompileActivationCheckpointing` | the bug above; verified to fail against the pre-fix pass |
| `test_unhandled_parallel_layer_raises` | a layer reached but not rewritten must stop the compile |
| `test_deferred_layer_run_eagerly_raises` | the eager-execution guard |

Existing tests in the file are unchanged and still pass (2x/4xH200, `-m sequential`: 6 passed).

The regression test was checked against the pre-fix pass to confirm it actually catches the bug:
with `iter_graphs`/the coverage check reverted and the new tests kept, it fails. The magnitude of
the divergence in that configuration (forward off by 2.7e-01) is from the standalone reproducer
above rather than the test itself, since the distributed-test harness reports only the worker exit
code.

Note that these tests, like all the AutoTP compile tests, are `@pytest.mark.sequential`, which
`tests/pytest.ini` excludes by default. They need the marker to run at all:

```bash
pytest tests/unit/compile/test_tp_compile.py -m sequential
```
### With this PR applied

The same comparison, run against pristine `master` and against this branch **in a single job on one
node**, so both arms see identical hardware. Ten Adam steps rather than one, so anything that drifts
rather than breaking outright has somewhere to show up. Reported as the worst per-step difference
against the module-level AutoTP path:

| config | | `master` (ba3246de) | **this PR** |
|---|---|---|---|
| activation checkpointing **on**, tp=2 | loss | 1.10e-01 **FAIL** | **0.00e+00 PASS** |
| | grad norm | 1.05e-01 | **0.00e+00** |
| activation checkpointing **on**, tp=4 | loss | 1.42e-01 **FAIL** | **1.19e-07 PASS** |
| | grad norm | 9.67e-02 | **5.96e-08** |
| activation checkpointing **off**, tp=2 | loss | 0.00e+00 PASS | 0.00e+00 PASS |
| | grad norm | 0.00e+00 | 0.00e+00 |

### CODE REVIEW

File: deepspeed/compile/passes/tp_compile.py [hunk 1]
```diff
@@ -141,8 +241,8 @@ def pass_canonicalize(gm: GraphModule, real_inputs):
 ]
 
 
-def apply_autotp(gm: GraphModule, real_inputs, passes=None):
+def apply_autotp(gm: GraphModule, real_inputs, passes=None, deferred_names=None):
     """Apply the AutoTP transformation passes to the graph."""
     for opt_pass in passes or AUTOTP_PASSES:
-        opt_pass(gm, real_inputs)
+        opt_pass(gm, real_inputs, deferred_names=deferred_names)
```
[the remark this answers is not in the document]
Comment (pengdurice) [reply deepspeed/compile/passes/tp_compile.py]: good catch, fixed.

**Review by tohtana — APPROVED** [review]
Thank you for the fix, @pengdurice!
This looks good to me.
